README.md
# isortd

[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/urm8/isortd/build?style=for-the-badge)](https://github.com/urm8/isortd/actions)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/isortd?style=for-the-badge)](https://pypi.org/project/isortd/)
[![PyPI](https://img.shields.io/pypi/v/isortd?style=for-the-badge)](https://pypi.org/project/isortd/)
[![Docker Image Version (latest by date)](https://img.shields.io/docker/v/urm8/isortd?style=for-the-badge)](https://hub.docker.com/repository/docker/urm8/isortd)

Simple http handler for [isort](https://github.com/PyCQA/isort) util. I liked the idea of putting
[black[d]](https://black.readthedocs.io/en/stable/blackd.html) into my docker compose file and using
[BlackConnect](https://plugins.jetbrains.com/plugin/14321-blackconnect) plugin for auto sort without setting up my dev
env every time, but I was still missing sort formatting tool, that would work the same way. So its here... Mb I'll
release [IsortConnect](https://github.com/urm8/IsortConnect) and it will be more usable.

## install

```
poetry add isortd
```

or

```
python -m pip install --upgrade isortd
```

## usage

I'd suggest you to run this [docker image](https://hub.docker.com/repository/docker/urm8/isortd) with something like:

```
docker run -d --name isortd --publish "47393:47393" urm8/isortd:latest
```

or just add it to your local docker-compose file \_0_/

## todo

* socket support
* some kind of filechangelistener daemon runnable from directory

CHANGELOG.md
# [unreleased]
## added
- makefile
- keepachangelog CHANGELOG.md
## fixed
- ci now pushes new versions to pypi
## changed
- .gitignore now ignores local .env file
## removed

tests/test_isortd.py
from isortd import __version__


def test_version():
    assert __version__ == "0.1.0"


isortd/main.py
from __future__ import annotations

import logging
import tempfile
from concurrent import futures
from functools import lru_cache
from pathlib import Path
from typing import Iterable

import aiohttp_cors
import click
from aiohttp import web
from isort import __version__ as isort_ver
from isort import code, settings
from isort.exceptions import ISortError

from isortd import __version__ as ver


@click.command(context_settings={"help_option_names": ["-h --help"]})
@click.option("--host", type=str, help="App host", default="localhost")
@click.option("--port", type=int, help="App port", default=47393)
def main(host, port):
    logging.basicConfig(level=logging.INFO)
    with futures.ProcessPoolExecutor() as executor:
        app = factory(executor)
        app.logger.info(f"isortd v{ver} (isort core v{isort_ver}) listening on {host} port {port}")
        web.run_app(app, host=host, port=port, handle_signals=True) or 0
    return 0


def factory(executor: futures.ProcessPoolExecutor) -> web.Application:
    app = web.Application()
    cors = aiohttp_cors.setup(app)
    handler = HttpHandler(executor)
    sort_resource = cors.add(app.router.add_resource("/"))
    cors.add(
        sort_resource.add_route("POST", handler.handle),
        {
            "*": aiohttp_cors.ResourceOptions(
                expose_headers="*",
                allow_headers=(
                    "Content-Type",
                    "Accept-Encoding",
                    "Content-Encoding",
                    "X-*",
                ),
            ),
        },
    )

    ping_resource = cors.add(app.router.add_resource("/ping"))
    cors.add(
        ping_resource.add_route("GET", pong),
        {
            "*": aiohttp_cors.ResourceOptions(
                allow_methods=["GET"],
                allow_credentials=True,
                expose_headers="*",
                allow_headers="*",
            ),
        },
    )
    return app


async def pong(*_):
    return web.Response(text=f"pong", status=200)


class HttpHandler:
    def __init__(self, executor: futures.ProcessPoolExecutor):
        self.executor = executor

    async def handle(self, request: web.Request):
        in_ = await request.text()
        try:
            fp = request.headers.get("XX-PATH")
            src = tuple(request.headers.get("XX-SRC", "").split(","))
            args = _parse_arguments(request.headers.items())
            cfg = _get_config(args, src)
        except ISortError as e:
            return web.Response(body=f"Failed to parse config: {e}", status=400)
        out = code(code=in_, config=cfg, file_path=Path(fp) if fp else None)
        if out:
            return web.Response(
                text=out,
                content_type=request.content_type,
                charset=request.charset,
                zlib_executor_size=1024,
            )
        return web.Response(status=201)


def _parse_arguments(headers: Iterable[tuple[str, str]]) -> tuple[str, ...]:
    normalized = tuple(
        sorted(
            f"{_normalize_headers(key)}={value}"
            for key, value in headers
            if key.startswith("X-")
        )
    )
    return normalized


def _normalize_headers(key: str):
    return key.lower().replace("x-", "")


@lru_cache(maxsize=128)
def _get_config(args: tuple[str, ...], src: list[str]):
    kwargs = {}
    if args:
        kwargs["settings_file"] = _write_temp_config(args)
    if src:
        kwargs["src_paths"] = src
    return settings.Config(**kwargs)


def _write_temp_config(args):
    with tempfile.NamedTemporaryFile("w", suffix=".toml", delete=False) as tmp:
        tmp.write("\n".join(("[tool.isort]", *args)))
        file_path = tmp.name
    return file_path


isortd/daemon.py
import os
from io import StringIO

from aiofile import AIOFile
from isort import settings
from loguru import logger
from watchgod import PythonWatcher, awatch

from isortd.isort_handler import IsortHandler

CUR_DIR = "."
PYPROJECT_TOML = "pyproject.toml"
ISORT_CFG = ".isort.cfg"
POSSIBLE_CFGS = [PYPROJECT_TOML, ISORT_CFG]


async def main():
    cfg = None
    curr_dir_contents = os.listdir(CUR_DIR)
    if not set(curr_dir_contents).intersection(POSSIBLE_CFGS):
        logger.warning("was not able to find any config file for isort daemon")
    else:
        cfg, *_ = set(POSSIBLE_CFGS).intersection(curr_dir_contents)

    if cfg is None:
        logger.warning("failed to find config file will use default settings")
    handler = IsortHandler(cfg=settings.Config(cfg))
    logger.info(f"start watching: {CUR_DIR}")
    async for changes in awatch(CUR_DIR, watcher_cls=PythonWatcher, debounce=100):
        for _, file_changed in changes:
            await try_isort(file_changed, handler)


async def try_isort(file_changed: str, handler: IsortHandler):
    try:
        fout = StringIO()
        with AIOFile(file_changed, "r") as fin:
            in_ = await fin.read()
            if out := handler.handle(fin):
                async with AIOFile(file_changed, "w") as wfout:
                    await wfout.write(fout.getvalue())
    except Exception as e:
        logger.error(f"shit happened: {e}")


if __name__ == "__main__":
    import asyncio

    asyncio.run(main())


isortd/http.py
from __future__ import annotations

import logging
import tempfile
from concurrent import futures
from functools import lru_cache
from pathlib import Path
from typing import Iterable

import aiohttp_cors
import click
from aiohttp import web
from isort import __version__ as isort_ver
from isort import code, settings
from isort.exceptions import ISortError

from isortd import __version__ as ver


@click.command(context_settings={"help_option_names": ["-h --help"]})
@click.option("--host", type=str, help="App host", default="localhost")
@click.option("--port", type=int, help="App port", default=47393)
def main(host, port):
    logging.basicConfig(level=logging.INFO)
    with futures.ProcessPoolExecutor() as executor:
        app = factory(executor)
        app.logger.info(
            f"isortd v{ver} (isort core v{isort_ver}) listening on {host} port {port}"
        )
        web.run_app(app, host=host, port=port, handle_signals=True) or 0
    return 0


def factory(executor: futures.ProcessPoolExecutor) -> web.Application:
    app = web.Application()
    cors = aiohttp_cors.setup(app)
    handler = HttpHandler(executor)
    sort_resource = cors.add(app.router.add_resource("/"))
    cors.add(
        sort_resource.add_route("POST", handler.handle),
        {
            "*": aiohttp_cors.ResourceOptions(
                expose_headers="*",
                allow_headers=(
                    "Content-Type",
                    "Accept-Encoding",
                    "Content-Encoding",
                    "X-*",
                ),
            ),
        },
    )

    ping_resource = cors.add(app.router.add_resource("/ping"))
    cors.add(
        ping_resource.add_route("GET", pong),
        {
            "*": aiohttp_cors.ResourceOptions(
                allow_methods=["GET"],
                allow_credentials=True,
                expose_headers="*",
                allow_headers="*",
            ),
        },
    )
    return app


async def pong(*_):
    return web.Response(text=f"pong", status=200)


class HttpHandler:
    def __init__(self, executor: futures.ProcessPoolExecutor):
        self.executor = executor

    async def handle(self, request: web.Request):
        in_ = await request.text()
        try:
            fp = request.headers.get("XX-PATH")
            src = tuple(request.headers.get("XX-SRC", "").split(","))
            args = _parse_arguments(request.headers.items())
            cfg = _get_config(args, src)
        except ISortError as e:
            return web.Response(body=f"Failed to parse config: {e}", status=400)
        out = code(code=in_, config=cfg, file_path=Path(fp) if fp else None)
        if out:
            return web.Response(
                text=out,
                content_type=request.content_type,
                charset=request.charset,
                zlib_executor_size=1024,
            )
        return web.Response(status=201)


def _parse_arguments(headers: Iterable[tuple[str, str]]) -> tuple[str, ...]:
    normalized = tuple(
        sorted(
            f"{_normalize_headers(key)}={value}"
            for key, value in headers
            if key.startswith("X-")
        )
    )
    return normalized


def _normalize_headers(key: str):
    return key.lower().replace("x-", "")


@lru_cache(maxsize=128)
def _get_config(args: tuple[str, ...], src: list[str]):
    kwargs = {}
    if args:
        kwargs["settings_file"] = _write_temp_config(args)
    if src:
        kwargs["src_paths"] = src
    return settings.Config(**kwargs)


def _write_temp_config(args):
    with tempfile.NamedTemporaryFile("w", suffix=".toml", delete=False) as tmp:
        tmp.write("\n".join(("[tool.isort]", *args)))
        file_path = tmp.name
    return file_path
